tee-worker/core/direct-rpc-server/src/rpc_responder.rs

```diff
@@ -226,9 +239,51 @@ where
 			.withdraw(&old_hash)
 			.ok_or(DirectRpcError::InvalidConnectionHash)?;
 
+		if !force_wait {
+			warn!("Unexpected force_wait = false for hash: {:?}", old_hash);
+		}
+
+		let mut result = RpcReturnValue::from_hex(&rpc_response.result)
+			.map_err(|e| DirectRpcError::Other(format!("{:?}", e).into()))?;
+
+		let mut new_response = rpc_response.clone();
+
 		// leave `rpc_response` untouched - it should be overwritten later anyway and keep on force waiting
 		self.connection_registry
 			.store(new_hash, connection_token, rpc_response, force_wait);
+
+		// it denotes a predecessor "streamed" hash whose status is either "TopExecuted" or "InSidechainBlock",
+		// if that's the case, don't put it back
+		match result.status {
+			DirectRequestStatus::TrustedOperationStatus(top_status, h) => {
+				if is_removable_top_status(&top_status) {
+					debug!("connection with old hash {:?} will be removed", old_hash);
+				} else {
+					// the swap comes a bit earlier, we store the old_hash back and mark it as "removable"
+					// it will be deleted when it reaches `is_removable_top_status`
+					result.status = DirectRequestStatus::TrustedOperationStatus(
+						TrustedOperationStatus::SuccessorExecuted,
+						h,
+					);
+					new_response.result = result.to_hex();
+					self.connection_registry.store(
+						old_hash,
+						connection_token,
+						new_response,
+						force_wait,
+					);
+				}
+			},
+			_ => {
+				self.connection_registry.store(
+					old_hash,
```

Comment:
why do we need to put back the `old_hash` for the other cases?

Reply:
Here other case means status other than `TrustedOperationStatus`, so it's either `Ok` or `Invalid`. In either case we don't want to touch it (at least not within swap_hash)